Inconsistent Behaviour of `node.removeChild()`

Hello,

I'm seeing some weird behaviour when calling `node.removeChild()`. If the xml source is "minified" (no pretty formatting) then it removes the first child from the node that I'm calling `removeChild()` on. If I call `removeChild()` on the dom instance itself, it replaces the dom instance with an empty string.

If however, the file is pretty printed, and I call `dom.removeChild()` or `node.removeChild()` and give it a child node (any child node in the tree) I get the tree returned back with just that node removed.

I'm using xmldom version 0.3.0 in nodejs version 14.12.0 and 10.16.0.

Here's a small mocha test that demonstrates what i'm talking about


```javascript
const { DOMParser } = require('xmldom');
const { expect } = require('chai');

const MINIFIED_XML = `<?xml version="1.0" encoding="UTF-8" standalone="yes"?><parent><child><grandchild attr="value"/></child></parent>`;
const MINIFIED_XML_WITHOUT_GRANDCHILD = `<?xml version="1.0" encoding="UTF-8" standalone="yes"?><parent><child/></parent>`;

const PRETTY_PRINTED_XML = `
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<parent>
  <child>
    <grandchild attr="value"/>
  </child>
</parent>
`.trim();
const PRETTY_PRINTED_XML_WITHOUT_GRANDCHILD = `
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<parent>
  <child>
    
  </child>
</parent>`.trim();

describe('remove child test', () => {
  it('removes all content from dom if minified', async () => {
    const dom = new DOMParser().parseFromString(MINIFIED_XML);
    const grandchildren = dom.getElementsByTagName('grandchild');

    expect(grandchildren.length).to.equal(1);
    expect(dom.toString()).to.equal(MINIFIED_XML);

    const grandchild = dom.removeChild(grandchildren[0]);

    expect(grandchild.toString()).to.equal('<grandchild attr="value"/>');
    expect(dom.toString()).to.equal(MINIFIED_XML_WITHOUT_GRANDCHILD);
    expect(dom.toString()).not.to.equal('');
  });

  it('correctly removes grandchild from dom if called on parent element', async () => {
    const dom = new DOMParser().parseFromString(MINIFIED_XML);
    const children = dom.getElementsByTagName('child');
    const grandchildren = dom.getElementsByTagName('grandchild');

    expect(children.length).to.equal(1);
    expect(grandchildren.length).to.equal(1);
    expect(dom.toString()).to.equal(MINIFIED_XML);

    const grandchild = children[0].removeChild(grandchildren[0]);

    expect(grandchild.toString()).to.equal('<grandchild attr="value"/>');
    expect(dom.toString()).to.equal(MINIFIED_XML_WITHOUT_GRANDCHILD);
    expect(dom.toString()).not.to.equal('');
  });

  it('removes just the grandchild node if pretty formatted', async () => {
    const dom = new DOMParser().parseFromString(PRETTY_PRINTED_XML);
    const grandchildren = dom.getElementsByTagName('grandchild');

    expect(grandchildren.length).to.equal(1);
    expect(dom.toString()).to.equal(PRETTY_PRINTED_XML);

    const grandchild = dom.removeChild(grandchildren[0]);

    expect(grandchild.toString()).to.equal('<grandchild attr="value"/>');
    expect(dom.toString()).not.to.equal('');
    expect(dom.toString()).to.equal(PRETTY_PRINTED_XML_WITHOUT_GRANDCHILD);
  });
});

```

Looking at the docs on MDN (https://developer.mozilla.org/en-US/docs/Web/API/Node/removeChild) I'd expect calling `removeChild` on the dom instance itself to throw a `TypeError` since I don't think the dom instance is supposed to have `removeChild` as a function? I'd also maybe expect that calling `removeChild()` on anything but the child's direct parent to throw a `NotFoundError` because the node isn't a child of the node.

However, I really like the behaviour that I'm seeing in the pretty printed version of the file. It's real handy to be able to call `removeChild` on the dom instance and get it to find and remove that child from the tree, wherever it is.

What is the expected behaviour of this method supposed to be?

Thanks!
Mike

Hey! Thanks for the quick response to the issue, greatly appreciated!

Yeah, I'm not sure what approach I'm looking for here, haha. We've got a work around in our code for this right now, but this bug certainly is a tricky one. For what it's worth, my vote is for strict adherence to the spec (as I understand it... I could be wrong):
  - `node.removeChild()` removes a child node from the tree and returns it iff the node to be removed is a child of the node `removeChild()` is being called on.
  - If the child node is not a child of the parent node, then throw a `NotFoundError` `DOMException`
  - If the child node doesn't exist in the DOM throw a `TypeError`

Though you're right. That'll be a doozy of a breaking change for many people I think. Anyway, enough of my opinions on the matter. I'm sure you'll figure out what the best approach here is. 😁 

To answer your question, calling `removeChild()` in this way always does _something_, and I never get any exceptions. Though what happens is certainly unexpected.

```javascript
const { DOMParser } = require('xmldom');
const { expect } = require('chai');

const ISSUE_CHECK = `
<xml>
  <a>
    <x/>
  </a>
  <b>
    <y/>
  </b>
</xml>
`.trim();

const ISSUE_CHECK_WITHOUT_Y = `
<xml>
  <a>
    <x/>
  </a>
  <b>
    
  </b>
</xml>
`.trim();

const ISSUE_CHECK_MINIFIED = `<xml><a><x/></a><b><y/></b></xml>`;
const ISSUE_CHECK_WITHOUT_Y_MINIFIED = `<xml><a><x/></a><b/></xml>`;

describe('remove child test', () => {
  it('Remove child from non-parent node pretty printed', async () => {
    const dom = new DOMParser().parseFromString(ISSUE_CHECK);
    const ys = dom.getElementsByTagName('y');
    const as = dom.getElementsByTagName('a');

    expect(dom.toString()).to.equal(ISSUE_CHECK);

    const removedY = as[0].removeChild(ys[0]);

    expect(removedY.toString()).to.equal('<y/>');
    expect(dom.toString()).not.to.equal('');
    expect(dom.toString()).to.equal(ISSUE_CHECK_WITHOUT_Y);
  });

  it('Remove child from non-parent node minified', async () => {
    const dom = new DOMParser().parseFromString(ISSUE_CHECK_MINIFIED);
    const ys = dom.getElementsByTagName('y');
    const as = dom.getElementsByTagName('a');

    expect(dom.toString()).to.equal(ISSUE_CHECK_MINIFIED);

    const removedY = as[0].removeChild(ys[0]);

    expect(removedY.toString()).to.equal('<y/>');
    expect(dom.toString()).not.to.equal('');
    expect(dom.toString()).to.equal(ISSUE_CHECK_WITHOUT_Y_MINIFIED);
  });
});

```

So the first test passes for me. The document is "pretty printed" and I try to remove a child (`<y />`) from a node that is not it's parent (`<a />`).  I would expect this to throw an exception. What actually happens is that `<y />` is removed from the dom as if I had called `removeChild()` on `<b />`.

The second test fails for me, but not in the way I'd expect. I'm doing the same operation, but this time the document is "minified". What happens here is that the node `<x />` is removed from `<b />` even though I'm calling `removeChild()` on `<a />` with the child `<y />`; which is super weird. e.g. here is the result of the second test:

```
AssertionError: expected '<xml><a/><b><y/></b></xml>' to equal '<xml><a><x/></a><b/></xml>'
Expected :<xml><a><x/></a><b/></xml>
Actual   :<xml><a/><b><y/></b></xml>
```

Thanks again for looking in to this. I hope all of this has been helpful 😄

--- a/lib/dom.js
+++ b/lib/dom.js
@@ -952,22 +952,26 @@ function _onUpdateChild(doc, el, newChild) {
  * @private
  */
 function _removeChild(parentNode, child) {
-	var previous = child.previousSibling;
-	var next = child.nextSibling;
-	if (previous) {
-		previous.nextSibling = next;
-	} else {
-		parentNode.firstChild = next;
+	if (parentNode !== child.parentNode) {
+		throw new DOMException(NOT_FOUND_ERR, "child's parent is not parent");
 	}
-	if (next) {
-		next.previousSibling = previous;
+	//var index = parentNode.childNodes.
+	var oldPreviousSibling = child.previousSibling;
+	var oldNextSibling = child.nextSibling;
+	if (oldPreviousSibling) {
+		oldPreviousSibling.nextSibling = oldNextSibling;
 	} else {
-		parentNode.lastChild = previous;
+		parentNode.firstChild = oldNextSibling;
 	}
+	if (oldNextSibling) {
+		oldNextSibling.previousSibling = oldPreviousSibling;
+	} else {
+		parentNode.lastChild = oldPreviousSibling;
+	}
+	_onUpdateChild(parentNode.ownerDocument, parentNode);
 	child.parentNode = null;
 	child.previousSibling = null;
 	child.nextSibling = null;
-	_onUpdateChild(parentNode.ownerDocument, parentNode);
 	return child;
 }
 
@@ -1345,10 +1349,11 @@ Document.prototype = {
 		return newChild;
 	},
 	removeChild: function (oldChild) {
-		if (this.documentElement == oldChild) {
+		var removed = _removeChild(this, oldChild);
+		if (removed === this.documentElement) {
 			this.documentElement = null;
 		}
-		return _removeChild(this, oldChild);
+		return removed;
 	},
 	replaceChild: function (newChild, oldChild) {
 		//raises

--- a/test/dom/document.test.js
+++ b/test/dom/document.test.js
@@ -379,8 +379,21 @@ describe('Document.prototype', () => {
 			// expect(doc.documentElement).toBeNull();
 			expect(initialElement.parentNode).toBeNull();
 			expect(initialElement.nextSibling).toBeNull();
-			// expect(initialElement.previousSibling).toBeNull();
+			expect(initialElement.previousSibling).toBeNull();
 			expect(doc.childNodes).toHaveLength(1);
+		});
+
+		it('Remove child from non-parent node throws', async () => {
+			const ISSUE_CHECK = `<xml>
+				<a><x/></a>
+				<b><y/></b>
+			</xml>`;
+			const dom = new DOMParser().parseFromString(ISSUE_CHECK);
+			const ys = dom.getElementsByTagName('y');
+			const as = dom.getElementsByTagName('a');
+
+			expect(() => as[0].removeChild(ys[0])).toThrow(DOMException);
+			expect(dom.toString()).toBe(ISSUE_CHECK);
 		});
 	});
 });

